### PR TITLE
Use ValueKeccak as the basis for Keccak

### PR DESCRIPTION
**Pull request by benaadams**
## Changes

- Use `ValueKeccak` as the basis for `Keccak` rather than `byte[]`
- Drop `KeccakStructRef` and use `ref readonly ValueKeccak` instead
- Drop `KeccakKey` and use `ValueKeccak` instead

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes Modified existing ones
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.Abi.Test/AbiTests.cs [hunk 1]
```diff
@@ -367,7 +367,7 @@ public void Test_single_address_no_signature(AbiEncodingStyle encodingStyle)
         [TestCase(AbiEncodingStyle.None)]
         public void Test_packed(AbiEncodingStyle encodingStyle)
         {
-            Keccak assetId = Keccak.Compute("assetId");
+            Keccak assetId = new Keccak(Keccak.Compute("assetId"));
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Abi.Test/AbiTests.cs:370]: Why not keep the old signature?

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs [hunk 2]
```diff
@@ -261,19 +204,36 @@ public class Keccak : IEquatable<Keccak>, IComparable<Keccak>
         /// </summary>
         public static Keccak MaxValue { get; } = new("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
 
-        public byte[] Bytes { get; }
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:264]: I was thinking of changing this to return a `ReadOnlySpan`. That would've significantly reduce the amount of change.

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs [hunk 3]
```diff
@@ -85,12 +84,12 @@ public ValueKeccak(ReadOnlySpan<byte> bytes)
...
             if (bytes.Length == 0)
             {
-                Bytes = OfAnEmptyString.Bytes;
+                _bytes = OfAnEmptyString._bytes;
                 return;
             }
 
             Debug.Assert(bytes.Length == ValueKeccak.MemorySize);
-            Bytes = Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(bytes));
+            _bytes = Unsafe.As<byte, Vector256<byte>>(ref MemoryMarshal.GetReference(bytes));
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:92]: Getting a bit confused here, to clarify, this do copy the bytes right? Just in an optimized way.

**Comment by asdacap** [comment]
Can you reduce the amount of change? This is too much to review. You also changed other struct to use `ValueKeccak`, can you split it up to a different PR?

File: src/Nethermind/Nethermind.Abi.Test/AbiTests.cs (hunk 1 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Abi.Test/AbiTests.cs]: Some places don't need the actual `Keccak` object if they then immediately call another method on it like `.ToString`

However reverted the signature and changed those places to use `ValueKeccak.Compute` instead

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs (hunk 2 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:264]: Changed `Span` to `Bytes`

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs (hunk 3 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:92]: Yes; or rather reinterprets to a `Vector256<byte>` and then assigns to the `Vector256<byte> _bytes` field.

Is a 256 bit / 32 byte copy as a single instruction

File: src/Nethermind/Nethermind.Blockchain.Test/BlockTreeTests.cs [hunk 4]
```diff
@@ -230,7 +230,7 @@ public void Cleans_invalid_blocks_before_starting()
             tree.SuggestBlock(block2);
             tree.SuggestBlock(block3);
 
-            blockInfosDb.Set(BlockTree.DeletePointerAddressInDb, block1.Hash.Bytes);
+            blockInfosDb.Set(BlockTree.DeletePointerAddressInDb, block1.Hash.ValueKeccak);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain.Test/BlockTreeTests.cs:233]: Hmm.. should be an unnecessary change. Just need to change `DbExtensions` to accept `ReadOnlySpan<byte>`.

File: src/Nethermind/Nethermind.Blockchain/Receipts/KeccaksIterator.cs [hunk 5]
```diff
@@ -20,17 +30,17 @@ public KeccaksIterator(Span<byte> data)
             Index = -1;
         }
 
-        public bool TryGetNext(out KeccakStructRef current)
+        public bool TryGetNext(out KeccakIteratorRef current)
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Blockchain/Receipts/KeccaksIterator.cs:33]: Hmm... why? Is it because of removal of `KeccakStructRef`?

File: src/Nethermind/Nethermind.Core/TransactionReceipt.cs [hunk 6]
```diff
@@ -58,8 +58,8 @@ public class TxReceipt
         /// </summary>
         public byte StatusCode { get; set; }
         public long BlockNumber { get; set; }
-        public KeccakStructRef BlockHash;
-        public KeccakStructRef TxHash;
+        public ref readonly ValueKeccak BlockHash;
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/TransactionReceipt.cs:61]: Interesting, does this mean it only store pointer to the value? Is this only valid for ref struct?

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/Eth/EthRpcModuleTests.cs [hunk 7]
```diff
@@ -369,7 +369,7 @@ public async Task Eth_get_balance_internal_error()
     public async Task Eth_get_balance_incorrect_parameters()
     {
         using Context ctx = await Context.Create();
-        string serialized = ctx.Test.TestEthRpc("eth_getBalance", TestItem.KeccakA.Bytes.ToHexString(true), "0x01");
+        string serialized = ctx.Test.TestEthRpc("eth_getBalance", TestItem.KeccakA.ToByteArray().ToHexString(true), "0x01");
```
Comment (asdacap) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/Eth/EthRpcModuleTests.cs:372]: Can't it `ToHexString` from `Span`?

File: src/Nethermind/Nethermind.Synchronization.Test/SnapSync/ProgressTrackerTests.cs [hunk 8]
```diff
@@ -58,26 +58,26 @@ public void Will_create_multiple_get_address_range_request()
 
         (SnapSyncBatch request, bool finished) = progressTracker.GetNextRequest();
         request.AccountRangeRequest.Should().NotBeNull();
-        request.AccountRangeRequest!.StartingHash.Bytes[0].Should().Be(0);
-        request.AccountRangeRequest.LimitHash!.Bytes[0].Should().Be(64);
+        request.AccountRangeRequest!.StartingHash.ToByteArray()[0].Should().Be(0);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Synchronization.Test/SnapSync/ProgressTrackerTests.cs:61]: NP: Unnecessary conversion.

File: src/Nethermind/Nethermind.Synchronization/Nethermind.Synchronization.csproj [hunk 9]
```diff
@@ -6,6 +6,9 @@
     
     
     <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    
+    
+    <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Synchronization/Nethermind.Synchronization.csproj:11]: Does this allow this for the whole project or just `unsafe` blocks?

File: src/Nethermind/Nethermind.Blockchain/Receipts/KeccaksIterator.cs (hunk 5 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Blockchain/Receipts/KeccaksIterator.cs:33]: `readonly ref` as a parameter is `in`, can't pass it `out` :(

So need so pass in another structure that can pass `out`

File: src/Nethermind/Nethermind.Core/TransactionReceipt.cs (hunk 6 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/TransactionReceipt.cs:61]: Only valid for `ref struct` can do `ref readonly` for a `readonly struct` and perhaps more confucingly can do `readonly ref readonly` for one that can only be set in the constructor (for example in a `readonly ref struct`)

File: src/Nethermind/Nethermind.Wallet/IWallet.cs [hunk 10]
```diff
@@ -16,7 +16,9 @@ public interface IWallet
         bool LockAccount(Address address);
         bool IsUnlocked(Address address);
         Signature Sign(Keccak message, Address address, SecureString passphrase = null);
+        Signature Sign(ValueKeccak message, Address address, SecureString passphrase = null);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Wallet/IWallet.cs:19]: Can we just keep the `ValueKeccak` one?

File: src/Nethermind/Nethermind.Core/TransactionReceipt.cs (hunk 6 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/TransactionReceipt.cs:61]: Yes, is pointer (ref) to value

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs [hunk 11]
```diff
@@ -101,30 +111,30 @@ public static ValueKeccak Compute(ReadOnlySpan<byte> input)
                 return OfAnEmptyString;
             }
 
-            ValueKeccak result = default;
-            KeccakHash.ComputeHashBytesToSpan(input, result.BytesAsSpan);
-            return result;
+            Span<byte> span = stackalloc byte[MemorySize];
+            KeccakHash.ComputeHashBytesToSpan(input, span);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:115]: Would it make any difference to create a `ValueKeccak` first, expose a `Span<byte>` (not read only), and use that as destination instead?

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs [hunk 12]
```diff
@@ -162,94 +177,33 @@ public string ToString(bool withZeroX)
...
-
     [DebuggerStepThrough]
+    [DebuggerDisplay("{ToString()}")]
     public class Keccak : IEquatable<Keccak>, IComparable<Keccak>
     {
         public const int Size = 32;
 
         public const int MemorySize =
-            MemorySizes.SmallObjectOverhead +
+            MemorySizes.SmallObjectOverhead -
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:187]: Why does it become `-`?

Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:187]: `SmallObjectOverhead` is `24`; but it only uses `16`

8 bytes object header, 8 bytes method table, no padding, 32 byte struct
```
Type layout for 'Keccak'
Size: 32 bytes. Paddings: 0 bytes (%0 of empty space)
|=========================================|
| Object Header (8 bytes)                 |
|-----------------------------------------|
| Method Table Ptr (8 bytes)              |
|=========================================|
|  0-31: ValueKeccak Bytes (32 bytes)     |
| |=====================================| |
| |  0-31: Vector256`1 Bytes (32 bytes) | |
| | |=============================|     | |
| | |   0-7: UInt64 _00 (8 bytes) |     | |
| | |-----------------------------|     | |
| | |  8-15: UInt64 _01 (8 bytes) |     | |
| | |-----------------------------|     | |
| | | 16-23: UInt64 _02 (8 bytes) |     | |
| | |-----------------------------|     | |
| | | 24-31: UInt64 _03 (8 bytes) |     | |
| | |=============================|     | |
| |=====================================| |
|=========================================|
```

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs (hunk 11 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:115]: Tried that; it got very inconstant in actually updating the `ValueKeccak`. I assume since its both a `readonly` struct and a 32byte register type the Jit makes assumptions and applies optimizations which means it doesn't always write back

Comment (asdacap) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:115]: Noted, can you add a comment on this?

**Review by asdacap — APPROVED** [review]
LGTM.

File: src/Nethermind/Nethermind.Synchronization/Nethermind.Synchronization.csproj (hunk 9 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Synchronization/Nethermind.Synchronization.csproj:7]: Whole project but only used it for `[SkipLocalsInit]` in the method where I also called `.Clear` on the stackalloc'd span

File: src/Nethermind/Nethermind.Wallet/IWallet.cs (hunk 10 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Wallet/IWallet.cs:18]: Removed other two

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs (hunk 11 above)
Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:115]: 🤔 I was thinking of a different place where modified after construction.

Let me take a look at this one again

Comment (benaadams) [reply src/Nethermind/Nethermind.Core/Crypto/Keccak.cs]: Looks good!

**Review by LukaszRozmej — COMMENTED** [review]
1. All methods should take ValueKeccak by ref
2. Dictionaries could use https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-7.0 ?
3. Can we remove Keccak class? (Potentially renaming ValueKeccak to just Keccak)

**Comment by benaadams** [comment]
> 3. Can we remove Keccak class? (Potentially renaming ValueKeccak to just Keccak)

Not without an even a larger refactor as it would break a lot of the Json outputs and `null` `Keccak`s are currently used for indicators in a lot of places and `ValueKeccak` is non-nullable (and `ValueKeccak?` is a poor substitute)

**Comment by benaadams** [comment]
> 1. All methods should take ValueKeccak by ref

Added; also made `Keccak.ValueKeccak` return a `readonly ref` that can be passed via `in` rather than going via implicit casting which will create a new copy

> 2. Dictionaries could use https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-7.0 ?

Can't use this; it returns the `KeyValuePair` by reference however the key to get it is still by value

> 3. Can we remove Keccak class? (Potentially renaming ValueKeccak to just Keccak)

Not without an even a larger refactor as it would break a lot of the Json outputs and null Keccaks are currently used for indicators in a lot of places and ValueKeccak is non-nullable (and ValueKeccak? is a poor substitute)

**Comment by LukaszRozmej** [comment]
> > 1. All methods should take ValueKeccak by ref
> 
> Added; also made `Keccak.ValueKeccak` return a `readonly ref` that can be passed via `in` rather than going via implicit casting which will create a new copy
> 
> > 2. Dictionaries could use https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.collectionsmarshal.getvaluereforadddefault?view=net-7.0 ?
> 
> Can't use this; it returns the `KeyValuePair` by reference however the key to get it is still by value
> 
> > 3. Can we remove Keccak class? (Potentially renaming ValueKeccak to just Keccak)
> 
> Not without an even a larger refactor as it would break a lot of the Json outputs and null Keccaks are currently used for indicators in a lot of places and ValueKeccak is non-nullable (and ValueKeccak? is a poor substitute)

at least I got 1 out of 3 right :D

File: src/Nethermind/Nethermind.Core/Crypto/Keccak.cs [hunk 13]
```diff
@@ -385,169 +339,43 @@ public override int GetHashCode()
...
+            if (k2 is null) return true;
 
-        public static bool operator !=(Keccak a, KeccakStructRef b)
-        {
-            return !(a == b);
+            return k1._keccak <= k2._keccak;
         }
 
-        public static bool operator !=(KeccakStructRef a, KeccakStructRef b)
+        public byte[] ToByteArray()
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Core/Crypto/Keccak.cs:376]: Please review each usage of this method. I think a lot can be changed to Span and avoid alloc. Or avoid it in other way.

File: src/Nethermind/Nethermind.Crypto/IEcdsa.cs [hunk 14]
```diff
@@ -8,7 +8,10 @@ namespace Nethermind.Crypto
     public interface IEcdsa
     {
         Signature Sign(PrivateKey privateKey, Keccak message);
+        Signature Sign(PrivateKey privateKey, in ValueKeccak message);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Crypto/IEcdsa.cs:11]: if we can get ValueKeccak out of Keccak, maybe we should remove Keccak overloads in all the classes/interfaces? Maybe add implicit cast to ValueKeccak?

File: src/Nethermind/Nethermind.TxPool/HashCache.cs [hunk 15]
```diff
@@ -35,28 +35,28 @@ internal class HashCache
...
         {
-            _longTermCache.Delete(hash);
+            _longTermCache.Delete(hash.ValueKeccak);
         }
 
         public void Delete(Keccak hash)
         {
-            _longTermCache.Delete(hash);
-            _currentBlockCache.Delete(hash);
+            _longTermCache.Delete(hash.ValueKeccak);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/HashCache.cs:58]: in ?

File: src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs [hunk 16]
```diff
@@ -106,7 +106,7 @@ private void StartBroadcast(Transaction tx)
         {
             NotifyPeersAboutLocalTx(tx);
             if (tx.Hash is not null)
-                _persistentTxs.TryInsert(tx.Hash, tx);
+                _persistentTxs.TryInsert(tx.Hash.ValueKeccak, tx);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxBroadcaster.cs:109]: in ?

File: src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs [hunk 17]
```diff
@@ -109,7 +109,7 @@ private void UpdateGroup(Address groupKey, Account groupValue, EnhancedSortedSet
             ReadOnlySpan<Transaction> txs = CollectionsMarshal.AsSpan(_transactionsToRemove);
             for (int i = 0; i < txs.Length; i++)
             {
-                TryRemove(txs[i].Hash!);
+                TryRemove(txs[i].Hash!.ValueKeccak);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs:112]: in ?

File: src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs [hunk 18]
```diff
@@ -95,7 +95,7 @@ private void UpdateGroup(Address groupKey, Account groupValue, EnhancedSortedSet
                     tx.GasBottleneck = changedGasBottleneck;
                     if (reAdd)
                     {
-                        _worstSortedValues.Add(tx, tx.Hash!);
+                        _worstSortedValues.Add(tx, tx.Hash!.ValueKeccak);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs:98]: in ?

File: src/Nethermind/Nethermind.TxPool.Test/Collections/DistinctValueSortedPoolTests.cs [hunk 19]
```diff
@@ -96,7 +96,7 @@ public void Same_transactions_are_all_replaced_with_highest_gas_price(bool gasPr
 
             foreach (var transaction in transactions)
             {
-                pool.TryInsert(transaction.Hash, transaction);
+                pool.TryInsert(transaction.Hash.ValueKeccak, transaction);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool.Test/Collections/DistinctValueSortedPoolTests.cs:99]: in ?

File: src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs [hunk 20]
```diff
@@ -42,7 +42,7 @@ public void SaveInCache(TrieNode node)
 
             public TrieNode FindCachedOrUnknown(Keccak hash)
             {
-                if (_objectsCache.TryGetValue(hash, out TrieNode trieNode))
+                if (_objectsCache.TryGetValue(hash.ValueKeccak, out TrieNode trieNode))
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs:45]: in ?

File: src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs [hunk 21]
```diff
@@ -49,7 +49,7 @@ protected override void UpdateGroup(Address groupKey, EnhancedSortedSet<Transact
                     change(tx);
                     if (reAdd)
                     {
-                        _worstSortedValues.Add(tx, tx.Hash!);
+                        _worstSortedValues.Add(tx, tx.Hash!.ValueKeccak);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/Collections/TxDistinctSortedPool.cs:52]: in ?

File: src/Nethermind/Nethermind.TxPool/TxPool.cs [hunk 22]
```diff
@@ -348,7 +348,7 @@ private AcceptTxResult AddCore(Transaction tx, TxFilteringState state, bool isPe
                 bool eip1559Enabled = _specProvider.GetCurrentHeadSpec().IsEip1559Enabled;
 
                 tx.GasBottleneck = tx.CalculateEffectiveGasPrice(eip1559Enabled, _headInfo.CurrentBaseFee);
-                bool inserted = _transactions.TryInsert(tx.Hash!, tx, out Transaction? removed);
+                bool inserted = _transactions.TryInsert(tx.Hash!.ValueKeccak, tx, out Transaction? removed);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.TxPool/TxPool.cs:351]: in?

**Review by LukaszRozmej — COMMENTED** [review]
Optimize `ToArray` usage.
Next steps: Introduce ValueAddress, ValueBloom ect.

**Comment by asdacap** [comment]
Can we expedite this. This will conflict with other optimization.

**Comment by benaadams** [comment]
A lot merged in https://github.com/NethermindEth/nethermind/pull/5715 can revisit the extras another time
